protobuf/algorithm.proto

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright (c) 2020, Arm Limited, All Rights Reserved
+ * SPDX-License-Identifier: Apache-2.0
+ */
+syntax = "proto3";
+
+package algorithm;
+
+message Algorithm {
+  message Hash {
+    enum Algorithm {
+      NONE = 0;
+      MD2 = 1; // DEPRECATED
+      MD4 = 2; // DEPRECATED
+      MD5 = 3; // DEPRECATED
+      RIPEMD160 = 4;
+      SHA_1 = 5; // DEPRECATED
+      SHA_224 = 6;
+      SHA_256 = 7;
+      SHA_384 = 8;
+      SHA_512 = 9;
+      SHA_512_224 = 10;
+      SHA_512_256 = 11;
+      SHA3_224 = 12;
+      SHA3_256 = 13;
+      SHA3_384 = 14;
+      SHA3_512 = 15;
+      ANY_HASH = 2047;
+    }
+
+    Algorithm algorithm = 1;
+  }
+  message Mac {
+    message Normal {
+      message Hmac { Hash hash_alg = 1; }
+      message CbcMac {}
+      message Cmac {}
+
+      oneof variant {
+        Hmac hmac = 1;
+        CbcMac cbc_mac = 2;
+        Cmac cmac = 3;
+      }
+    }
+    message Truncated {
+      Normal mac_alg = 1;
+      uint32 mac_length = 2;
+    }
+
+    oneof variant {
+      Normal normal = 1;
+      Truncated truncated = 2;
+    }
+  }
+  message Cipher {
+    enum Algorithm {
+      NONE = 0;
+      STREAM_CIPHER = 1;
+      CTR = 2;
+      CFB = 3;
+      OFB = 4;
+      XTS = 5;
+      ECB_NO_PADDING = 6;
+      CBC_NO_PADDING = 7;
+      CBC_PKCS7 = 8;
+    }
+
+    Algorithm algorithm = 1;
+  }
+  message Aead {
+    message Normal {
+      enum Algorithm {
+        NONE = 0;
+        CCM = 1;
+        GCM = 2;
+        CHACHA20_POLY1305 = 3;
+      }
+
+      Algorithm algorithm = 1;
+    }
+    message AeadWithShortenedTag {
+      Normal aead_alg = 1;
+      uint32 tag_length = 2;
+    }
+
+    oneof variant {
+      Normal normal = 1;
+      AeadWithShortenedTag aead_with_shortened_tag = 2;
+    }
+  }
+  message AsymmetricSignature {
+    message RsaPkcs1v15Sign { Hash hash_alg = 1; }
+    message RsaPkcs1v15SignRaw {}
+    message RsaPss { Hash hash_alg = 1; }
+    message Ecdsa { Hash hash_alg = 1; }
+    message EcdsaAny {}
```

Comment:
Good catch. I would argue though to keep it as it has a reserved value in the spec? The `NONE` value of `hash_alg` is more, in my opinion, something that we do because of protobuf than because it is in the spec. I could not find a valid hash algorithm value in the spec that means "No hash".

Reply:
True, but then again, the spec doesn't really use enums! Might sound a bit pedantic, but in that case should we have some sort of note next to the `NONE` variant to mark that it should not be used explicitly because it's essentially invalid?

Reply:
Yes that is a good point and when parsing the enum value from protobuf generated code to the interface language, activate the encoding error case if the value is equal to `NONE`.

Reply:
There's actually [a `[deprecated = true]` option](https://developers.google.com/protocol-buffers/docs/proto3#options) for fields but I don't know if it applies to enum variants (I'd say probably not)